### PR TITLE
YPE-642 - verse action popover — highlights, copy & share 

### PR DESCRIPTION
**Pull request by cameronapak**
[VIDEO OVERVIEW](https://cap.so/s/s8m4wp273mh8jxm)

[YPE-642](https://lifechurch.atlassian.net/browse/YPE-642)

Adds a verse action popover to `BibleReader`: tap verse(s) to highlight, copy, or share.

## What's in it
- **Highlights** — 5 iOS-matched colors, painted as a 35%-opacity fill. Tap a color to apply, tap its X to remove.
- **Copy / Share** — bible.com-style output: curly-quoted text (verse numbers + footnote markers stripped), blank line, then `Book Ch:vv VERSION`. Non-contiguous picks join with ` ... `; ranges collapse (`1-3`). Share uses the Web Share API, clipboard fallback.
- **Persistence** — localStorage only, keyed by `bible_id` + verse USFM to mirror the future `highlight` API shape. **No network this PR** (server sync is a separate ticket).

## Behavior worth a look
- **Sticky bar** — when the anchored verse scrolls off, the bar docks to the edge it exited (down → top, up → bottom) so the actions stay reachable.
- **Fades in place** — applying a highlight or clicking away fades the bar out where it sits (no anchor-loss jump).
- No auth gating; highlights render regardless of sign-in. No cross-chapter selection.

## Notes
- Footnote markers darken to `--yv-muted-foreground` over a highlight for AA contrast.
- Design rationale + ADRs: `docs/adr/YPE-642-verse-action-popover.md`.

## Testing
- 164 unit tests pass (incl. new `verse-action-popover` + `verse-share` suites); typecheck + lint clean.
- Storybook: **Components/BibleTextView → VerseSelection**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[YPE-642]: https://lifechurch.atlassian.net/browse/YPE-642?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<h3>Greptile Summary</h3>

This PR adds a full verse-action popover to `BibleReader`: tap verse(s) to reveal a floating bar with five highlight colors (localStorage-persisted), Copy, and Share. It includes the sticky/docked bar when the anchored verse scrolls off-screen, a "frozen snapshot" fade-out animation, and a clean `verse-share` utility layer.

- **New components**: `VerseActionPopover` (Radix `PopoverAnchor virtualRef`, IntersectionObserver docking), `verse-share.ts` (formatting helpers), two icon components; wired into `BibleTextView` with `selectedVerses` / `highlightedVerses` / `onVerseSelect`.
- **BibleReader API additions**: `onCopy` / `onShare` override props expose the structured `BibleReaderShareData` payload so React Native / Expo hosts can forward selections over the native bridge.
- **Persistence**: highlights keyed as `youversion-platform:highlights:<versionId>` → `{ "<USFM>": "<hex>" }` in localStorage; version-switch clears the store synchronously during render to prevent stale highlights flashing over new text.

<h3>Confidence Score: 5/5</h3>

Safe to merge — no functional regressions or data-integrity issues found in the new verse-action popover, highlight persistence, or copy/share flows.

The core feature implementation is sound: highlight state is correctly scoped per version, the synchronous store-clear during render prevents stale flashes, the frozen-snapshot animation pattern handles the anchor-loss edge case, and the copy/share formatting utility is well-tested. The only findings are a misleading test comment, a theoretical Concurrent Mode edge case on a ref mutation during render, and a silent bad-CSS path when localStorage contains a malformed hex. None affect correctness under normal usage.

packages/ui/src/components/verse-action-popover.test.tsx — AC5a comment overstates which colours are hidden; packages/ui/src/components/verse-action-popover.tsx — frozen-view ref mutation during render; packages/ui/src/components/verse.tsx — hexToRgba lacks a length guard.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/ui/src/components/verse-action-popover.tsx | New component: Radix-based floating action bar with IntersectionObserver docking, frozen-snapshot fade-out, and color circle controls. Well-structured. The prefers-reduced-motion and non-unique aria-label issues from previous review cycles remain unaddressed. |
| packages/ui/src/components/bible-reader.tsx | Adds selection + highlight state, localStorage load/persist, and popover wiring to Content. Synchronous store-clear during render (React adjust-state-during-render pattern) correctly prevents stale-version flash. |
| packages/ui/src/lib/verse-share.ts | Clean pure-function formatting layer: formatVerseNumbers, joinVerseTexts, buildVerseReference, buildVerseShareText. Well-tested; edge cases (empty version abbreviation, non-contiguous verses) handled correctly. |
| packages/ui/src/components/verse.tsx | Adds hexToRgba + getCleanVerseText helpers, updates useLayoutEffect to paint inline backgroundColor fills, and wires isHighlighted to VerseFootnoteButton for AA contrast on highlight fills. |
| packages/ui/src/components/verse-action-popover.test.tsx | 470-line test file covering all 9 ACs. AC5a test comment misstates expected circle count (says '3 apply' but implementation and AC7 test both confirm 5 apply circles appear when multiple active colors). Assertions pass but the comment is misleading for future maintainers. |
| packages/ui/src/lib/verse-share.test.ts | Good coverage of formatVerseNumbers (ranges, non-contiguous, dedup), joinVerseTexts (gaps, trimming), buildVerseReference, and buildVerseShareText (bible.com format, empty version). |
| packages/ui/src/styles/global.css | Adds .yv-v-selected underline rule in yv-sdk-bible-reader layer. Correct scoping via [data-yv-sdk] prefix. |
| packages/ui/src/components/verse.stories.tsx | VerseSelectionDemo updated to drive the real VerseActionPopover instead of manual Highlight/Clear buttons. Mirrors the bible-reader.tsx wiring faithfully. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant User
    participant BibleTextView
    participant Content as BibleReader.Content
    participant VAP as VerseActionPopover
    participant LS as localStorage

    User->>BibleTextView: Tap verse N
    BibleTextView->>Content: onVerseSelect([N])
    Content->>Content: handleVerseSelect([N])
    Content->>Content: "querySelector(.yv-v[v=N]) → anchorElement"
    Content->>VAP: "open=true, anchorElement, selectedVerses=[N]"

    VAP->>VAP: IntersectionObserver observes anchorElement
    Note over VAP: Verse scrolls off screen
    VAP->>VAP: setDockEdge(top or bottom)
    VAP->>VAP: Switches to virtual docked anchor

    User->>VAP: Tap highlight color
    VAP->>Content: onHighlight(color)
    Content->>Content: handleHighlight(color)
    Content->>LS: localStorage.setItem(key, JSON)
    Content->>Content: closeAndClearSelection()
    Content->>VAP: "open=false (frozen snapshot animates out)"

    User->>VAP: Tap Copy
    VAP->>Content: onCopy()
    Content->>Content: buildSelectionShareData()
    Content->>BibleTextView: getCleanVerseText(container, verse)
    Content->>Content: buildVerseShareText(...)
    alt onCopy prop provided
        Content->>User: onCopy(BibleReaderShareData)
    else
        Content->>User: navigator.clipboard.writeText(text)
    end
    Content->>Content: closeAndClearSelection()
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant User
    participant BibleTextView
    participant Content as BibleReader.Content
    participant VAP as VerseActionPopover
    participant LS as localStorage

    User->>BibleTextView: Tap verse N
    BibleTextView->>Content: onVerseSelect([N])
    Content->>Content: handleVerseSelect([N])
    Content->>Content: "querySelector(.yv-v[v=N]) → anchorElement"
    Content->>VAP: "open=true, anchorElement, selectedVerses=[N]"

    VAP->>VAP: IntersectionObserver observes anchorElement
    Note over VAP: Verse scrolls off screen
    VAP->>VAP: setDockEdge(top or bottom)
    VAP->>VAP: Switches to virtual docked anchor

    User->>VAP: Tap highlight color
    VAP->>Content: onHighlight(color)
    Content->>Content: handleHighlight(color)
    Content->>LS: localStorage.setItem(key, JSON)
    Content->>Content: closeAndClearSelection()
    Content->>VAP: "open=false (frozen snapshot animates out)"

    User->>VAP: Tap Copy
    VAP->>Content: onCopy()
    Content->>Content: buildSelectionShareData()
    Content->>BibleTextView: getCleanVerseText(container, verse)
    Content->>Content: buildVerseShareText(...)
    alt onCopy prop provided
        Content->>User: onCopy(BibleReaderShareData)
    else
        Content->>User: navigator.clipboard.writeText(text)
    end
    Content->>Content: closeAndClearSelection()
```

</a>

<details open><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `packages/ui/src/components/verse-action-popover.tsx`, line 1275-1282 ([link](https://github.com/youversion/platform-sdk-react/blob/3e7a7c1ff24329953f2c9d5afc1afb3df6f92be2/packages/ui/src/components/verse-action-popover.tsx#L1275-L1282)) 

   <a href="#">P1</a> **Non-unique aria-labels on color circle buttons**

   All five "apply" buttons receive the same label (`"Apply highlight"`) and all "clear" buttons receive the same label (`"Clear highlight"`). A screen reader user navigating the toolbar hears the same announcement for every color swatch and has no way to know which color they are about to apply or remove. This blocks the highlight-by-color feature for keyboard/AT users entirely.

   The i18n keys `applyHighlightAriaLabel` and `clearHighlightAriaLabel` are also locked to a bare string in all three locale files, so there is no placeholder available for the color name. A fix requires adding a color-name map (e.g. `{ fffe00: 'yellow', '5dff79': 'green', ... }`) and updating the i18n keys to accept an interpolation variable (e.g. `"Apply {{color}} highlight"` / `"Clear {{color}} highlight"`).

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: packages/ui/src/components/verse-action-popover.tsx
   Line: 1275-1282

   Comment:
   **Non-unique aria-labels on color circle buttons**

   All five "apply" buttons receive the same label (`"Apply highlight"`) and all "clear" buttons receive the same label (`"Clear highlight"`). A screen reader user navigating the toolbar hears the same announcement for every color swatch and has no way to know which color they are about to apply or remove. This blocks the highlight-by-color feature for keyboard/AT users entirely.

   The i18n keys `applyHighlightAriaLabel` and `clearHighlightAriaLabel` are also locked to a bare string in all three locale files, so there is no placeholder available for the color name. A fix requires adding a color-name map (e.g. `{ fffe00: 'yellow', '5dff79': 'green', ... }`) and updating the i18n keys to accept an interpolation variable (e.g. `"Apply {{color}} highlight"` / `"Clear {{color}} highlight"`).

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fui%2Fsrc%2Fcomponents%2Fverse-action-popover.tsx%0ALine%3A%201275-1282%0A%0AComment%3A%0A**Non-unique%20aria-labels%20on%20color%20circle%20buttons**%0A%0AAll%20five%20%22apply%22%20buttons%20receive%20the%20same%20label%20%28%60%22Apply%20highlight%22%60%29%20and%20all%20%22clear%22%20buttons%20receive%20the%20same%20label%20%28%60%22Clear%20highlight%22%60%29.%20A%20screen%20reader%20user%20navigating%20the%20toolbar%20hears%20the%20same%20announcement%20for%20every%20color%20swatch%20and%20has%20no%20way%20to%20know%20which%20color%20they%20are%20about%20to%20apply%20or%20remove.%20This%20blocks%20the%20highlight-by-color%20feature%20for%20keyboard%2FAT%20users%20entirely.%0A%0AThe%20i18n%20keys%20%60applyHighlightAriaLabel%60%20and%20%60clearHighlightAriaLabel%60%20are%20also%20locked%20to%20a%20bare%20string%20in%20all%20three%20locale%20files%2C%20so%20there%20is%20no%20placeholder%20available%20for%20the%20color%20name.%20A%20fix%20requires%20adding%20a%20color-name%20map%20%28e.g.%20%60%7B%20fffe00%3A%20'yellow'%2C%20'5dff79'%3A%20'green'%2C%20...%20%7D%60%29%20and%20updating%20the%20i18n%20keys%20to%20accept%20an%20interpolation%20variable%20%28e.g.%20%60%22Apply%20%7B%7Bcolor%7D%7D%20highlight%22%60%20%2F%20%60%22Clear%20%7B%7Bcolor%7D%7D%20highlight%22%60%29.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youversion%2Fplatform-sdk-react&pr=269&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fui%2Fsrc%2Fcomponents%2Fverse-action-popover.tsx%0ALine%3A%201275-1282%0A%0AComment%3A%0A**Non-unique%20aria-labels%20on%20color%20circle%20buttons**%0A%0AAll%20five%20%22apply%22%20buttons%20receive%20the%20same%20label%20%28%60%22Apply%20highlight%22%60%29%20and%20all%20%22clear%22%20buttons%20receive%20the%20same%20label%20%28%60%22Clear%20highlight%22%60%29.%20A%20screen%20reader%20user%20navigating%20the%20toolbar%20hears%20the%20same%20announcement%20for%20every%20color%20swatch%20and%20has%20no%20way%20to%20know%20which%20color%20they%20are%20about%20to%20apply%20or%20remove.%20This%20blocks%20the%20highlight-by-color%20feature%20for%20keyboard%2FAT%20users%20entirely.%0A%0AThe%20i18n%20keys%20%60applyHighlightAriaLabel%60%20and%20%60clearHighlightAriaLabel%60%20are%20also%20locked%20to%20a%20bare%20string%20in%20all%20three%20locale%20files%2C%20so%20there%20is%20no%20placeholder%20available%20for%20the%20color%20name.%20A%20fix%20requires%20adding%20a%20color-name%20map%20%28e.g.%20%60%7B%20fffe00%3A%20'yellow'%2C%20'5dff79'%3A%20'green'%2C%20...%20%7D%60%29%20and%20updating%20the%20i18n%20keys%20to%20accept%20an%20interpolation%20variable%20%28e.g.%20%60%22Apply%20%7B%7Bcolor%7D%7D%20highlight%22%60%20%2F%20%60%22Clear%20%7B%7Bcolor%7D%7D%20highlight%22%60%29.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=269&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22youversion%2Fplatform-sdk-react%22%20on%20the%20existing%20branch%20%22cp%2FYPE-642-react-sdk-bible-reader-ui-verse-action-popover-with-highlight-colors-copy-and-share%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22cp%2FYPE-642-react-sdk-bible-reader-ui-verse-action-popover-with-highlight-colors-copy-and-share%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fui%2Fsrc%2Fcomponents%2Fverse-action-popover.tsx%0ALine%3A%201275-1282%0A%0AComment%3A%0A**Non-unique%20aria-labels%20on%20color%20circle%20buttons**%0A%0AAll%20five%20%22apply%22%20buttons%20receive%20the%20same%20label%20%28%60%22Apply%20highlight%22%60%29%20and%20all%20%22clear%22%20buttons%20receive%20the%20same%20label%20%28%60%22Clear%20highlight%22%60%29.%20A%20screen%20reader%20user%20navigating%20the%20toolbar%20hears%20the%20same%20announcement%20for%20every%20color%20swatch%20and%20has%20no%20way%20to%20know%20which%20color%20they%20are%20about%20to%20apply%20or%20remove.%20This%20blocks%20the%20highlight-by-color%20feature%20for%20keyboard%2FAT%20users%20entirely.%0A%0AThe%20i18n%20keys%20%60applyHighlightAriaLabel%60%20and%20%60clearHighlightAriaLabel%60%20are%20also%20locked%20to%20a%20bare%20string%20in%20all%20three%20locale%20files%2C%20so%20there%20is%20no%20placeholder%20available%20for%20the%20color%20name.%20A%20fix%20requires%20adding%20a%20color-name%20map%20%28e.g.%20%60%7B%20fffe00%3A%20'yellow'%2C%20'5dff79'%3A%20'green'%2C%20...%20%7D%60%29%20and%20updating%20the%20i18n%20keys%20to%20accept%20an%20interpolation%20variable%20%28e.g.%20%60%22Apply%20%7B%7Bcolor%7D%7D%20highlight%22%60%20%2F%20%60%22Clear%20%7B%7Bcolor%7D%7D%20highlight%22%60%29.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youversion%2Fplatform-sdk-react&pr=269&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>

2. `packages/ui/src/components/verse-action-popover.tsx`, line 1240-1244 ([link](https://github.com/youversion/platform-sdk-react/blob/3e7a7c1ff24329953f2c9d5afc1afb3df6f92be2/packages/ui/src/components/verse-action-popover.tsx#L1240-L1244)) 

   <a href="#">P2</a> **No `prefers-reduced-motion` guard on popover animations**

   The popover uses `fade-in-0`, `zoom-in-95`, `slide-in-from-top-2`, and related Tailwind animation utilities with a 180 ms custom duration. Users who prefer reduced motion (via the OS setting) will still see these transitions. The review guidelines call out respecting this media query for all animations.

   Adding `motion-safe:` variants (e.g. `yv:motion-safe:data-[state=open]:animate-in`) or a global `@media (prefers-reduced-motion: reduce)` override that sets `animation-duration: 0s` on `[data-yv-sdk]` elements would resolve this.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: packages/ui/src/components/verse-action-popover.tsx
   Line: 1240-1244

   Comment:
   **No `prefers-reduced-motion` guard on popover animations**

   The popover uses `fade-in-0`, `zoom-in-95`, `slide-in-from-top-2`, and related Tailwind animation utilities with a 180 ms custom duration. Users who prefer reduced motion (via the OS setting) will still see these transitions. The review guidelines call out respecting this media query for all animations.

   Adding `motion-safe:` variants (e.g. `yv:motion-safe:data-[state=open]:animate-in`) or a global `@media (prefers-reduced-motion: reduce)` override that sets `animation-duration: 0s` on `[data-yv-sdk]` elements would resolve this.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fui%2Fsrc%2Fcomponents%2Fverse-action-popover.tsx%0ALine%3A%201240-1244%0A%0AComment%3A%0A**No%20%60prefers-reduced-motion%60%20guard%20on%20popover%20animations**%0A%0AThe%20popover%20uses%20%60fade-in-0%60%2C%20%60zoom-in-95%60%2C%20%60slide-in-from-top-2%60%2C%20and%20related%20Tailwind%20animation%20utilities%20with%20a%20180%20ms%20custom%20duration.%20Users%20who%20prefer%20reduced%20motion%20%28via%20the%20OS%20setting%29%20will%20still%20see%20these%20transitions.%20The%20review%20guidelines%20call%20out%20respecting%20this%20media%20query%20for%20all%20animations.%0A%0AAdding%20%60motion-safe%3A%60%20variants%20%28e.g.%20%60yv%3Amotion-safe%3Adata-%5Bstate%3Dopen%5D%3Aanimate-in%60%29%20or%20a%20global%20%60%40media%20%28prefers-reduced-motion%3A%20reduce%29%60%20override%20that%20sets%20%60animation-duration%3A%200s%60%20on%20%60%5Bdata-yv-sdk%5D%60%20elements%20would%20resolve%20this.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youversion%2Fplatform-sdk-react&pr=269&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fui%2Fsrc%2Fcomponents%2Fverse-action-popover.tsx%0ALine%3A%201240-1244%0A%0AComment%3A%0A**No%20%60prefers-reduced-motion%60%20guard%20on%20popover%20animations**%0A%0AThe%20popover%20uses%20%60fade-in-0%60%2C%20%60zoom-in-95%60%2C%20%60slide-in-from-top-2%60%2C%20and%20related%20Tailwind%20animation%20utilities%20with%20a%20180%20ms%20custom%20duration.%20Users%20who%20prefer%20reduced%20motion%20%28via%20the%20OS%20setting%29%20will%20still%20see%20these%20transitions.%20The%20review%20guidelines%20call%20out%20respecting%20this%20media%20query%20for%20all%20animations.%0A%0AAdding%20%60motion-safe%3A%60%20variants%20%28e.g.%20%60yv%3Amotion-safe%3Adata-%5Bstate%3Dopen%5D%3Aanimate-in%60%29%20or%20a%20global%20%60%40media%20%28prefers-reduced-motion%3A%20reduce%29%60%20override%20that%20sets%20%60animation-duration%3A%200s%60%20on%20%60%5Bdata-yv-sdk%5D%60%20elements%20would%20resolve%20this.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=269&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22youversion%2Fplatform-sdk-react%22%20on%20the%20existing%20branch%20%22cp%2FYPE-642-react-sdk-bible-reader-ui-verse-action-popover-with-highlight-colors-copy-and-share%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22cp%2FYPE-642-react-sdk-bible-reader-ui-verse-action-popover-with-highlight-colors-copy-and-share%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fui%2Fsrc%2Fcomponents%2Fverse-action-popover.tsx%0ALine%3A%201240-1244%0A%0AComment%3A%0A**No%20%60prefers-reduced-motion%60%20guard%20on%20popover%20animations**%0A%0AThe%20popover%20uses%20%60fade-in-0%60%2C%20%60zoom-in-95%60%2C%20%60slide-in-from-top-2%60%2C%20and%20related%20Tailwind%20animation%20utilities%20with%20a%20180%20ms%20custom%20duration.%20Users%20who%20prefer%20reduced%20motion%20%28via%20the%20OS%20setting%29%20will%20still%20see%20these%20transitions.%20The%20review%20guidelines%20call%20out%20respecting%20this%20media%20query%20for%20all%20animations.%0A%0AAdding%20%60motion-safe%3A%60%20variants%20%28e.g.%20%60yv%3Amotion-safe%3Adata-%5Bstate%3Dopen%5D%3Aanimate-in%60%29%20or%20a%20global%20%60%40media%20%28prefers-reduced-motion%3A%20reduce%29%60%20override%20that%20sets%20%60animation-duration%3A%200s%60%20on%20%60%5Bdata-yv-sdk%5D%60%20elements%20would%20resolve%20this.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youversion%2Fplatform-sdk-react&pr=269&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>
</details>

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into cp/YPE-642-reac..."](https://github.com/youversion/platform-sdk-react/commit/bcf2ff88d8d3a71453de87712d911f5a7dd02a63) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39371341)</sub>

### CODE REVIEW

File: packages/ui/src/components/bible-reader.tsx [hunk 1]
```diff
@@ -462,6 +465,177 @@ function Content() {
...
+      chapter,
+      versionAbbreviation: version?.localized_abbreviation ?? '',
+    });
+  }
+
+  function handleCopy() {
+    const text = buildSelectionText();
+    if (text) void navigator.clipboard?.writeText(text);
+    closeAndClearSelection();
+  }
```
Comment (cameronapak) [inline packages/ui/src/components/bible-reader.tsx:681]: nitpick: We may need the ability to pass in an override function for React Native Expo SDK.

File: packages/ui/src/components/bible-reader.tsx [hunk 2]
```diff
@@ -462,6 +465,177 @@ function Content() {
...
+          // Cancelled or failed — keep the popover open (AC 4).
+        });
+      return;
+    }
+    // No Web Share support (e.g. most desktop browsers) — fall back to clipboard.
+    if (text && typeof navigator !== 'undefined') {
+      void navigator.clipboard?.writeText(text);
+    }
+    closeAndClearSelection();
+  }
```
Comment (cameronapak) [inline packages/ui/src/components/bible-reader.tsx:709]: nitpick: We may need the ability to pass in an override function for React Native Expo SDK.

File: packages/ui/src/components/bible-reader.tsx (hunk 1 above)
Comment (Dustin-Kelley) [reply packages/ui/src/components/bible-reader.tsx:681]: ahh yes this is true. Similar to how handle the share sheet?

File: .changeset/verse-action-popover.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@youversion/platform-react-ui": minor
+---
+
+Add a verse action popover to `BibleReader`. Tapping verses selects them (shown with an underline) and opens a popover anchored to the last-selected verse with five highlight colors, Copy, and Share. Highlights apply a translucent fill, persist to `localStorage` per Bible version (shaped like the future highlight API), and can be removed individually. Copy/Share output mirrors bible.com formatting: the verse text in curly quotes, gaps in a non-contiguous selection joined with `...`, followed by the `Book Chapter:verses VERSION` reference. Share uses the Web Share API and falls back to copying where it isn't available.
```
[the remark this answers is not in the document]
Comment (Dustin-Kelley) [reply .changeset/verse-action-popover.md:5]: hmm, not sure on this one. This seems too small to be a major bump?

File: packages/ui/src/components/bible-reader.tsx (hunk 1 above)
Comment (cameronapak) [reply packages/ui/src/components/bible-reader.tsx:681]: Yes!

File: packages/ui/src/components/bible-reader.tsx [hunk 4]
```diff
@@ -462,6 +465,177 @@ function Content() {
...
+  }, [highlightsStorageKey]);
+
+  // Navigating away (book/chapter/version) drops the selection — those verses no
+  // longer exist on screen (ADR-007).
+  useEffect(() => {
+    setSelectedVerses([]);
+    setPopoverOpen(false);
+    setAnchorElement(null);
+    lastSelectionRef.current = [];
+  }, [book, chapter, versionId]);
```
[the remark this answers is not in the document]
Comment (cameronapak) [reply packages/ui/src/components/bible-reader.tsx:548]: Fixed in 232ec26. Clearing inside the load effect still paints one stale frame (the effect runs after paint), so instead the store now resets synchronously during render when the version key changes (React's "adjust state during render" pattern) — the previous version's highlights can never paint over the new text.

File: packages/ui/src/components/bible-reader.tsx (hunk 1 above)
Comment (cameronapak) [reply packages/ui/src/components/bible-reader.tsx:681]: Done in 232ec26. Added optional `onCopy` / `onShare` props to `BibleReader.Root`. When provided they receive a structured `BibleReaderShareData` payload and suppress the default Web Share / clipboard flow, mirroring `VerseOfTheDay`'s `onShare` — so an Expo host can bridge to native clipboard/share.

File: packages/ui/src/components/bible-reader.tsx (hunk 2 above)
Comment (cameronapak) [reply packages/ui/src/components/bible-reader.tsx:709]: Covered by the same `onCopy` / `onShare` overrides added in 232ec26 (both Copy and Share now short-circuit to the host callbacks when supplied).

File: packages/ui/src/components/bible-reader.tsx [hunk 5]
```diff
@@ -462,6 +502,221 @@ function Content() {
     scrollContainerRef.current?.scrollTo({ top: 0 });
   }, [book, chapter]);
 
+  // ---- Verse selection + highlights ------------------------------------------
+  // Selection is ephemeral; highlights persist to localStorage only for now
```
Comment (Dustin-Kelley) [inline packages/ui/src/components/bible-reader.tsx:506]: question(non-blocking): So is this saying in the future highlights will persist only on the api and we are just using `localStorage` for now until the api is complete?

File: packages/ui/src/components/bible-reader.tsx [hunk 6]
```diff
@@ -462,6 +502,221 @@ function Content() {
...
+  useEffect(() => {
+    let data: Record<string, string> = {};
+    try {
+      const raw = localStorage.getItem(highlightsStorageKey);
+      if (raw) data = JSON.parse(raw) as Record<string, string>;
+    } catch {
+      // Ignore (unavailable or malformed storage).
+    }
+    setHighlightStore(data);
+  }, [highlightsStorageKey]);
```
Comment (Dustin-Kelley) [inline packages/ui/src/components/bible-reader.tsx:539]: praise: Nice use case for this effect 😎. Also just for fun check this out https://www.npmjs.com/package/use-local-storage. I've used it in the past and it makes things clean and nice. Not needed for this but thought I'd share

**Review by Dustin-Kelley — APPROVED** [review]
LGTM!

File: packages/ui/src/components/bible-reader.tsx (hunk 5 above)
Comment (cameronapak) [reply packages/ui/src/components/bible-reader.tsx:506]: Precisely!
